### ISSUE TITLE
Unable to attach running container on remote server

### ISSUE THREAD

**Issue by aska-0096** [microsoft/vscode-remote-release#6943]
- VSCode Version: 1.69.1
- Local OS Version: windows 10
- Remote OS Version: Ubuntu 18.04.6 LTS
- Remote Extension/Connection Type: SSH + Containers
- Logs:
- [2022-07-14T03:06:29.878Z] Start: Run: ssh -F C:\Users\haocwang\.ssh\config mi200 /bin/sh
[2022-07-14T03:06:29.878Z] Remote-Containers 0.241.3 in VS Code 1.69.1 (b06ae3b2d2dbfe28bca3134cc6be65935cdfea6a).
[2022-07-14T03:06:29.891Z] findLocalWindowsExecutable: Exectuable 'ssh' not found on PATH 'C:\Program Files (x86)\Google\Chrome\Application;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files\PuTTY\;C:\Program Files\Perforce;C:\Program Files\Git\cmd;C:\Program Files (x86)\Citrix\System32\;C:\Program Files\Citrix\Virtual Desktop Agent\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\110\Tools\Binn\;C:\Users\haocwang\AppData\Local\Microsoft\WindowsApps;;C:\Users\haocwang\AppData\Local\Programs\Microsoft VS Code\bin'.
[2022-07-14T03:06:29.894Z] Start: Run in host: id -un
[2022-07-14T03:06:29.896Z] Stop (18 ms): Run: ssh -F C:\Users\haocwang\.ssh\config mi200 /bin/sh
[2022-07-14T03:06:29.898Z] Stop (20 ms): Run: ssh -F C:\Users\haocwang\.ssh\config mi200 /bin/sh
[2022-07-14T03:07:11.460Z] Start: Run: ssh -F C:\Users\haocwang\.ssh\config mi200 /bin/sh
[2022-07-14T03:07:11.472Z] findLocalWindowsExecutable: Exectuable 'ssh' not found on PATH 'C:\Program Files (x86)\Google\Chrome\Application;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files\PuTTY\;C:\Program Files\Perforce;C:\Program Files\Git\cmd;C:\Program Files (x86)\Citrix\System32\;C:\Program Files\Citrix\Virtual Desktop Agent\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\110\Tools\Binn\;C:\Users\haocwang\AppData\Local\Microsoft\WindowsApps;;C:\Users\haocwang\AppData\Local\Programs\Microsoft VS Code\bin'.

Steps to Reproduce:

1. Connect to remote server using "RemoteSSH: connect to Host", password needed.
2. Right click one of running containers inside the docker plugin extension, choose "Attach visual studio code"
3. Failed with:
    Error running command remote-containers.attachToRunningContainerFromViewlet: Shell server failed: Error: spawn C:\Users\haocwang\AppData\Local\Programs\Microsoft VS Code\ssh ENOENT at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19) at onErrorNT (node:internal/child_process:477:16) at processTicksAndRejections (node:internal/process/task_queues:83:21). This is likely caused by the extension that contributes remote-containers.attachToRunningContainerFromViewlet.
4. Try to use F1 + RemoteContainer: attach to running container, similar error info Shell server failed appeared.

Basically this problem similar to #4396.

**Comment by chrmarti** [comment]
Where is you `ssh` executable located on your local machine? Does setting its absolute path in the user settings help?
<img width="725" alt="image" src="https://user-images.githubusercontent.com/9205389/178930453-eab52c43-b650-4646-8c9b-3b05d3a61ec6.png">

**Comment by aska-0096** [comment]
1. This local machine is a remote desktop using windows server OS, I don't have the administrator right to install openSSH, is there any other way to get ssh.exe?  
2. Why did RemoteContainer need ssh.exe on my local machine instead on the remote server?

**Comment by aska-0096** [comment]
> Where is you `ssh` executable located on your local machine? Does setting its absolute path in the user settings help? <img alt="image" width="725" src="https://user-images.githubusercontent.com/9205389/178930453-eab52c43-b650-4646-8c9b-3b05d3a61ec6.png">

I found a putty.exe, is it usable here or I should install a OpenSSH anyway.

**Comment by aska-0096** [comment]
Hi @chrmarti , any further comments on updated information?

**Comment by shimpossible** [comment]
fyi. Win10 supplies a copy of openssl under c:\windows\system32\OpenSSH

**Comment by matteverson** [comment]
I had a similar issue when I used the Remote SSH extension to connect to a remote server, then tried to attach to a container on that remote server. When using `c:\windows\system32\OpenSSH\ssh.exe` as remote.ssh.path:

```
[2022-08-22 22:44:40.610] [exthost] [trace] ExtHostCommands#$executeContributedCommand remote-containers.attachToRunningContainer
[2022-08-22 22:44:40.616] [exthost] [error] Shell server failed: Error: spawn c:\windows\system32\OpenSSH\ssh.exe ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
    at onErrorNT (node:internal/child_process:477:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) remote-containers.attachToRunningContainer {"value":"ms-vscode-remote.remote-containers","_lower":"ms-vscode-remote.remote-containers"}
```
When this value was not set at all:
```
[2022-08-22 22:41:58.592] [exthost] [error] Shell server failed: Error: spawn C:\Program Files (x86)\Microsoft VS Code\ssh ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
    at onErrorNT (node:internal/child_process:477:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) remote-containers.attachToRunningContainer {"value":"ms-vscode-remote.remote-containers","_lower":"ms-vscode-remote.remote-containers"}
```

I was able to work around the problem by checking the Remote SSH log to see what ssh binary it was using for the original connection to the remote server. In my case that was `C:\WINDOWS\Sysnative\OpenSSH\ssh.exe`:
```
[22:36:20.108] Checking ssh with "C:\WINDOWS\Sysnative\OpenSSH\ssh.exe -V"
[22:36:20.243] > O
[22:36:20.244] > penSSH_for_Windows_8.1p1, LibreSSL 3.0.2
```

Plugging that absolute path into the setting for remote.ssh.host allowed me to attach to the container or use the "Reopen in container" options.

**Comment by Xiaotian0726** [comment]
> Where is you `ssh` executable located on your local machine? Does setting its absolute path in the user settings help? <img alt="image" width="725" src="https://user-images.githubusercontent.com/9205389/178930453-eab52c43-b650-4646-8c9b-3b05d3a61ec6.png">

I've already set this value but the error is still there: 

![remotesshpath](https://user-images.githubusercontent.com/49309571/196661882-adc16743-86c5-4c50-b2a2-34a36aed5d0d.png)

![error](https://user-images.githubusercontent.com/49309571/196661944-75281c47-460e-47ec-991d-12d281367cec.png)

![infolder](https://user-images.githubusercontent.com/49309571/196661965-00ba01f2-1f6a-48b4-a9ec-f27fa3c85fbf.png)

**Comment by Xiaotian0726** [comment]
> > Where is you `ssh` executable located on your local machine? Does setting its absolute path in the user settings help? <img alt="image" width="725" src="https://user-images.githubusercontent.com/9205389/178930453-eab52c43-b650-4646-8c9b-3b05d3a61ec6.png">
> 
> I've already set this value but the error is still there:
> 
> ![remotesshpath](https://user-images.githubusercontent.com/49309571/196661882-adc16743-86c5-4c50-b2a2-34a36aed5d0d.png)
> 
> ![error](https://user-images.githubusercontent.com/49309571/196661944-75281c47-460e-47ec-991d-12d281367cec.png)
> 
> ![infolder](https://user-images.githubusercontent.com/49309571/196661965-00ba01f2-1f6a-48b4-a9ec-f27fa3c85fbf.png)

Solved.

![path](https://user-images.githubusercontent.com/49309571/196680551-096f32df-3127-41ac-8bc8-c9fa1f63800f.png)

**Comment by S-Mann** [comment]
> > > Where is you `ssh` executable located on your local machine? Does setting its absolute path in the user settings help? <img alt="image" width="725" src="https://user-images.githubusercontent.com/9205389/178930453-eab52c43-b650-4646-8c9b-3b05d3a61ec6.png">
> > 
> > 
> > I've already set this value but the error is still there:
> > ![remotesshpath](https://user-images.githubusercontent.com/49309571/196661882-adc16743-86c5-4c50-b2a2-34a36aed5d0d.png)
> > ![error](https://user-images.githubusercontent.com/49309571/196661944-75281c47-460e-47ec-991d-12d281367cec.png)
> > ![infolder](https://user-images.githubusercontent.com/49309571/196661965-00ba01f2-1f6a-48b4-a9ec-f27fa3c85fbf.png)
> 
> Solved.
> 
> ![path](https://user-images.githubusercontent.com/49309571/196680551-096f32df-3127-41ac-8bc8-c9fa1f63800f.png)

Best to use `ssh.exe`, since the ssh binary is in the windows path env var.
This is weirdly an issue specific to vscode running on windows trying to tunnel through a linux machine, I think it's the way vscode is trying to do SSH proxy tunneling, the proxy command should in theory work as `ssh` is recognized binary for both OS.

**Comment by bigdaronlee163** [comment]
![image](https://user-images.githubusercontent.com/49524138/201604391-5e8a5418-f2fe-4587-afad-3039cf6a4c25.png)
将vscode中的配置路径`C:\Windows\System32\OpenSSH\`改成绝对路径就好了。`C:\Windows\System32\OpenSSH\ssh.exe`

**Comment by chrmarti** [comment]
Make sure you use the full path to the ssh executable in the `Remote-SSH: Path` setting including the filename `ssh.exe` (not only the folder). That setting is not needed when `ssh` is on the `PATH`.
